File: dynamic-swap-program/prover/circuits/escrow_settle/escrow_settle.go
package escrow_settle

import (
	"github.com/consensys/gnark/frontend"
	"github.com/consensys/gnark/std/math/cmp"

	"zolana/prover/circuits/gadget"
	spp "zolana/prover/circuits/spp_transaction/shared"
)

const blindingSeedDomain = 0x44535458 // DSTX; matches the SDK's settle_blinding_seed.

// Circuit resolves an escrow -- settle or price-refund -- in a single circuit/VK
// so the resolving transaction never reveals which outcome occurred. The proof
// shape, account list, and verifying key are identical in both cases; the
// outcome is derived inside Define and is NOT observable, because MaxPrice is a
// private witness (bound to the order UTXO's DataHash) rather than a public
// input:
//
//	isSettle = (ExecutionPrice <= MaxPrice)
//
// Every escrow is priced at creation (commit is folded into create_escrow), so
// ExecutionPrice is always nonzero -- asserted below, an uncommitted order can
// never be proven. An order with an acceptable price settles; one whose price
// moved past MaxPrice refunds; settle and refund are indistinguishable on-chain.
//
// 2-in (order, reservation) / 3-out (recipient, maker_counter, maker_source), the
// exact IN2_OUT3 shape. There is no shared pool: the reservation input alone funds
// the recipient's payout and the maker's counter-asset change. On refund,
// MakerSource is a zero-amount output rather than an omitted one, so the shape
// does not differ between outcomes.
type Circuit struct {
	Public PublicInputs

	// Each UTXO carries the raw id of the tree it lives in as a sibling witness;
	// spp.UtxoHashCircuit folds it in as the second Poseidon element.
	OrderIn             spp.UtxoCircuitFields
	OrderInTreeID       frontend.Variable
	ReservationIn       spp.UtxoCircuitFields
	ReservationInTreeID frontend.Variable

	RecipientOut       spp.UtxoCircuitFields
	RecipientOutTreeID frontend.Variable
	MakerCounter       spp.UtxoCircuitFields
	MakerCounterTreeID frontend.Variable
	MakerSource        spp.UtxoCircuitFields
	MakerSourceTreeID  frontend.Variable

	OrderAmount frontend.Variable

	// RecipientOwnerHash, MaxPrice, and CreatedAt are private witnesses, bound to
	// OrderIn.DataHash (= Poseidon(RecipientOwnerHash, MaxPrice, CreatedAt), the
	// same commitment escrow_open wrote). Keeping MaxPrice private is what hides
	// settle-vs-refund: an observer knows the public ExecutionPrice but not the
	// threshold it is compared against. RecipientOwnerHash stays private too -- it
	// is the taker's owner-hash (= the source UTXO's owner escrow_open committed),
	// pinned by the public OrderInHash and bound to RecipientOut.Owner below, so
	// the payout destination is enforced without ever being revealed on-chain.
	RecipientOwnerHash frontend.Variable
	MaxPrice           frontend.Variable
	CreatedAt          frontend.Variable

	ExternalDataHash  frontend.Variable
	PrivateTxBlinding frontend.Variable
}

func (c *Circuit) Define(api frontend.API) error {
	orderInHash := c.checkOrderInputUtxo(api)
	// The escrow creator and settler both hold these openings. The first
	// nullifier is read from SPP instruction data by the native program, so the
	// settler cannot substitute a seed that prevents the recipient's recovery.
	blindingSeed := gadget.PoseidonHash(api, []frontend.Variable{blindingSeedDomain, c.OrderIn.Blinding, c.ReservationIn.Blinding})
	seed := spp.DeriveOutputBlindingSeed(api, c.Public.FirstNullifier, blindingSeed)
	api.AssertIsEqual(c.PrivateTxBlinding, spp.DerivePrivateTxBlinding(api, c.Public.FirstNullifier, blindingSeed))
	for i, blinding := range []frontend.Variable{c.RecipientOut.Blinding, c.MakerCounter.Blinding, c.MakerSource.Blinding} {
		api.AssertIsEqual(blinding, spp.DeriveOutputBlinding(api, c.Public.FirstNullifier, seed, i))
	}

	// Bind the private MaxPrice/CreatedAt to the order UTXO's committed DataHash
	// so the prover cannot choose a MaxPrice that flips the outcome. OrderInHash
	// is public and pins OrderIn's whole hash (incl. DataHash), so a false
	// MaxPrice cannot satisfy this equality.
	api.AssertIsEqual(c.OrderIn.DataHash, gadget.PoseidonHash(api, []frontend.Variable{
		c.RecipientOwnerHash,
		c.MaxPrice,
		c.CreatedAt,
	}))

	// Pin MaxPrice to 64 bits before the bounded comparator below: cmp
	// .IsLessOrEqual is only well-defined on in-range operands, so an out-of-
	// range MaxPrice would let the prover force isSettle either way. escrow_open
	// already bounds the value committed into DataHash, but this makes the
	// comparator's precondition locally explicit rather than cross-circuit.
	api.ToBinary(c.MaxPrice, 64)

	// Every escrow is priced at creation, so ExecutionPrice is always nonzero;
	// assert it so an uncommitted order can never be proven (rather than routing
	// a zero price to a free settle). The outcome is then purely the price
	// comparison against the private MaxPrice.
	api.AssertIsDifferent(c.Public.ExecutionPrice, 0)
	isSettle := cmp.IsLessOrEqual(api, c.Public.ExecutionPrice, c.MaxPrice)

	// Settle: owed = OrderAmount * ExecutionPrice, remainder = reserved - owed.
	// Refund: owed = 0, remainder = reserved (the full reservation credited back
	// to the maker). reserved (order_amount * max_price) is fixed at reservation
	// time.
	settleOwed := api.Mul(c.OrderAmount, c.Public.ExecutionPrice)
	owed := api.Select(isSettle, settleOwed, 0)
	reserved := api.Mul(c.OrderAmount, c.MaxPrice)
	remainder := api.Sub(reserved, owed)

	reservationInHash := c.checkReservationInputUtxo(api, reserved)

	// Settle: RecipientOut pays `owed` of the reservation's (destination) asset.
	// Refund: RecipientOut pays the full OrderAmount back in the order's (source)
	// asset. RecipientOwnerHash is identical either way.
	recipientAmount := api.Select(isSettle, owed, c.OrderAmount)
	recipientAsset := api.Select(isSettle, c.ReservationIn.Asset, c.OrderIn.Asset)
	recipientOutHash := c.checkRecipientOutputUtxo(api, recipientAmount, recipientAsset)

	// The maker's counter-asset leg: the unspent reservation (remainder) returns
	// to the maker. Replaces the old pool output -- there is no pool UTXO to add.
	makerCounterHash := c.checkMakerCounterOutputUtxo(api, remainder)

	// Settle: MakerSource receives OrderAmount of the settled source asset.
	// Refund: MakerSource is a zero-amount output -- present in both cases so the
	// shape never differs, but valueless when refunding.
	makerSourceAmount := api.Select(isSettle, c.OrderAmount, 0)
	makerSourceHash := c.checkMakerSourceOutputUtxo(api, makerSourceAmount)

	privateTxHashInputs{
		OrderInputUtxoHash:         orderInHash,
		ReservationInputUtxoHash:   reservationInHash,
		RecipientOutputUtxoHash:    recipientOutHash,
		MakerCounterOutputUtxoHash: makerCounterHash,
		MakerSourceOutputUtxoHash:  makerSourceHash,
		ExternalDataHash:           c.ExternalDataHash,
		PrivateTxBlinding:          c.PrivateTxBlinding,
		PrivateTxHash:              c.Public.PrivateTxHash,
	}.Check(api)

	c.Public.Check(api, orderInHash, reservationInHash)
	return nil
}

// PublicInputs folds PrivateTxHash and ExecutionPrice (the public pair price the
// native program reads from the escrow account) with OrderInHash and
// ReservationInHash (the witnessed input UTXOs' own reconstructed hashes,
// asserted equal in Check below), plus AuthorityOwnerHash (without which
// MakerCounter/MakerSource's Owner fields would be free witnesses a prover could
// redirect). MaxPrice and RecipientOwnerHash are deliberately NOT here -- both are
// private witnesses bound via OrderIn.DataHash: MaxPrice keeps the settle-vs-refund
// outcome hidden, and RecipientOwnerHash keeps the payout destination confidential
// (the public OrderInHash pins the DataHash, so neither can be forged). The native
// program recomputes this hash from on-chain state (`Escrow.execution_price`,
// `Escrow.escrow_utxo_hash`, `Escrow.reservation_utxo_hash`,
// `Pair.authority_owner_hash`).
type PublicInputs struct {
	PublicInputHash frontend.Variable `gnark:",public"`

	PrivateTxHash      frontend.Variable
	ExecutionPrice     frontend.Variable
	OrderInHash        frontend.Variable
	ReservationInHash  frontend.Variable
	AuthorityOwnerHash frontend.Variable
	FirstNullifier     frontend.Variable
}

func (p PublicInputs) Check(api frontend.API, orderInHash, reservationInHash frontend.Variable) {
	api.AssertIsEqual(p.OrderInHash, orderInHash)
	api.AssertIsEqual(p.ReservationInHash, reservationInHash)
	publicInputHash := gadget.PoseidonHash(api, []frontend.Variable{
		p.PrivateTxHash,
		p.ExecutionPrice,
		p.OrderInHash,
		p.ReservationInHash,
		p.AuthorityOwnerHash,
		p.FirstNullifier,
	})
	api.AssertIsEqual(p.PublicInputHash, publicInputHash)
}

type privateTxHashInputs struct {
	OrderInputUtxoHash         frontend.Variable
	ReservationInputUtxoHash   frontend.Variable
	RecipientOutputUtxoHash    frontend.Variable
	MakerCounterOutputUtxoHash frontend.Variable
	MakerSourceOutputUtxoHash  frontend.Variable
	ExternalDataHash           frontend.Variable
	PrivateTxBlinding          frontend.Variable
	PrivateTxHash              frontend.Variable
}

func (t privateTxHashInputs) Check(api frontend.API) {
	// 2-in/3-out; output order (recipient, maker_counter, maker_source) must match
	// the native program's output indices and the SDK.
	inputHashes := []frontend.Variable{
		t.OrderInputUtxoHash,
		t.ReservationInputUtxoHash,
	}
	outputHashes := []frontend.Variable{
		t.RecipientOutputUtxoHash,
		t.MakerCounterOutputUtxoHash,
		t.MakerSourceOutputUtxoHash,
	}
	addressHashes := []frontend.Variable{
		frontend.Variable(0),
		frontend.Variable(0),
	}

	privateTxHash := spp.PrivateTxHashCircuit(
		api,
		inputHashes,
		outputHashes,
		addressHashes,
		t.ExternalDataHash,
		t.PrivateTxBlinding,
	)
	api.AssertIsEqual(privateTxHash, t.PrivateTxHash)
}

func (c *Circuit) checkOrderInputUtxo(api frontend.API) frontend.Variable {
	api.AssertIsEqual(c.OrderIn.Domain, spp.UtxoDomain)
	api.AssertIsEqual(c.OrderIn.RingDataHash, 0)
	api.AssertIsEqual(c.OrderIn.RingProgramID, 0)
	api.AssertIsEqual(c.OrderIn.Amount, c.OrderAmount)
	return spp.UtxoHashCircuit(api, c.OrderIn, c.OrderInTreeID)
}

func (c *Circuit) checkReservationInputUtxo(api frontend.API, reserved frontend.Variable) frontend.Variable {
	api.AssertIsEqual(c.ReservationIn.Domain, spp.UtxoDomain)
	api.AssertIsEqual(c.ReservationIn.RingDataHash, 0)
	api.AssertIsEqual(c.ReservationIn.RingProgramID, 0)
	api.AssertIsEqual(c.ReservationIn.Amount, reserved)
	return spp.UtxoHashCircuit(api, c.ReservationIn, c.ReservationInTreeID)
}

func (c *Circuit) checkRecipientOutputUtxo(api frontend.API, amount, asset frontend.Variable) frontend.Variable {
	api.AssertIsEqual(c.RecipientOut.Domain, spp.UtxoDomain)
	api.AssertIsEqual(c.RecipientOut.RingDataHash, 0)
	api.AssertIsEqual(c.RecipientOut.RingProgramID, 0)
	api.AssertIsEqual(c.RecipientOut.DataHash, 0)
	api.AssertIsEqual(c.RecipientOut.Asset, asset)
	api.AssertIsEqual(c.RecipientOut.Amount, amount)
	api.AssertIsEqual(c.RecipientOut.Owner, c.RecipientOwnerHash)
	return spp.UtxoHashCircuit(api, c.RecipientOut, c.RecipientOutTreeID)
}

// checkMakerCounterOutputUtxo is the maker's counter-asset leg: the unspent
// reservation (remainder) returned to the maker's own note. It replaces the old
// pool output -- there is no pool UTXO, so the amount is exactly `remainder`, not
// pool_in + remainder. Asset is the reservation's (destination) asset.
func (c *Circuit) checkMakerCounterOutputUtxo(api frontend.API, remainder frontend.Variable) frontend.Variable {
	api.AssertIsEqual(c.MakerCounter.Domain, spp.UtxoDomain)
	api.AssertIsEqual(c.MakerCounter.RingDataHash, 0)
	api.AssertIsEqual(c.MakerCounter.RingProgramID, 0)
	api.AssertIsEqual(c.MakerCounter.DataHash, 0)
	api.AssertIsEqual(c.MakerCounter.Asset, c.ReservationIn.Asset)
	api.AssertIsEqual(c.MakerCounter.Owner, c.Public.AuthorityOwnerHash)

	api.AssertIsEqual(c.MakerCounter.Amount, remainder)
	api.ToBinary(c.MakerCounter.Amount, 64)

	return spp.UtxoHashCircuit(api, c.MakerCounter, c.MakerCounterTreeID)
}

// checkMakerSourceOutputUtxo is the pair authority's (maker's) own shielded UTXO
// receiving the settled source asset on a settle outcome. On a refund outcome
// `amount` is 0: the output is still produced (so the shape never differs between
// outcomes) but carries no value.
func (c *Circuit) checkMakerSourceOutputUtxo(api frontend.API, amount frontend.Variable) frontend.Variable {
	api.AssertIsEqual(c.MakerSource.Domain, spp.UtxoDomain)
	api.AssertIsEqual(c.MakerSource.RingDataHash, 0)
	api.AssertIsEqual(c.MakerSource.RingProgramID, 0)
	api.AssertIsEqual(c.MakerSource.DataHash, 0)
	api.AssertIsEqual(c.MakerSource.Asset, c.OrderIn.Asset)
	api.AssertIsEqual(c.MakerSource.Amount, amount)
	api.AssertIsEqual(c.MakerSource.Owner, c.Public.AuthorityOwnerHash)
	return spp.UtxoHashCircuit(api, c.MakerSource, c.MakerSourceTreeID)
}

File: dynamic-swap-program/prover/src/escrow_settle.rs
use std::collections::HashMap;

use crate::{
    bytes_to_decimal_string,
    ffi::{self, CircuitId},
    proof::{negate_and_compress_proof, OrderProof, ProofError},
    utxo::utxo_witness_entries,
    ProofInputUtxo,
};

/// Proof inputs for the `escrow_settle` circuit -- the single circuit `settle`
/// uses for both outcomes (settle and price-refund). Exact 2-in (order,
/// reservation) / 3-out (recipient, maker_counter, maker_source), no padding.
/// `max_price` and `created_at` are PRIVATE witnesses (bound to the order UTXO's
/// data hash), not public inputs -- keeping `max_price` private is what hides the
/// settle-vs-refund outcome. `execution_price` stays public (it is the public
/// pair price snapshot the escrow was priced at).
#[derive(Debug, Clone)]
pub struct EscrowSettleProofInputs {
    pub public_input_hash: [u8; 32],
    pub private_tx_hash: [u8; 32],
    pub first_nullifier: [u8; 32],
    pub execution_price: u64,
    /// Private witness, re-opened from the order UTXO's data hash in-circuit.
    pub max_price: u64,
    /// Private witness (the order's `created_at`), the third preimage of the
    /// order UTXO's data hash `Poseidon(recipient_owner_hash, max_price, created_at)`.
    pub created_at: u64,
    /// The order-input UTXO's own hash -- the escrow account's current
    /// on-chain `Escrow.escrow_utxo_hash`, asserted equal in-circuit to
    /// `Hash(order_in)`.
    pub order_in_hash: [u8; 32],
    /// The reservation-input UTXO's own hash -- `Escrow.reservation_utxo_hash`.
    pub reservation_in_hash: [u8; 32],
    /// Private witness: the taker's owner-hash (the source UTXO's owner that
    /// `escrow_open` committed into the order UTXO's DataHash), re-opened here and
    /// bound to `RecipientOut.Owner`. Pinned by the public `OrderInHash`.
    pub recipient_owner_hash: [u8; 32],
    /// Owner-hash bound to `MakerCounter.Owner`/`MakerSource.Owner` --
    /// `Pair.authority_owner_hash`.
    pub authority_owner_hash: [u8; 32],
    pub order_amount: u64,
    pub order_in: ProofInputUtxo,
    pub reservation_in: ProofInputUtxo,
    pub recipient_out: ProofInputUtxo,
    pub maker_counter: ProofInputUtxo,
    pub maker_source: ProofInputUtxo,
    pub external_data_hash: [u8; 32],
    pub private_tx_blinding: [u8; 32],
}

impl EscrowSettleProofInputs {
    fn witness(&self) -> ffi::WitnessMap {
        let mut map = HashMap::new();
        map.insert(
            "Public_PublicInputHash".to_string(),
            vec![bytes_to_decimal_string(&self.public_input_hash)],
        );
        map.insert(
            "Public_PrivateTxHash".to_string(),
            vec![bytes_to_decimal_string(&self.private_tx_hash)],
        );
        map.insert(
            "Public_FirstNullifier".to_string(),
            vec![bytes_to_decimal_string(&self.first_nullifier)],
        );
        map.insert(
            "Public_ExecutionPrice".to_string(),
            vec![self.execution_price.to_string()],
        );
        map.insert(
            "Public_OrderInHash".to_string(),
            vec![bytes_to_decimal_string(&self.order_in_hash)],
        );
        map.insert(
            "Public_ReservationInHash".to_string(),
            vec![bytes_to_decimal_string(&self.reservation_in_hash)],
        );
        map.insert(
            "Public_AuthorityOwnerHash".to_string(),
            vec![bytes_to_decimal_string(&self.authority_owner_hash)],
        );
        // Private witness (bound to the order UTXO's DataHash, pinned by the public
        // OrderInHash), so its key carries no `Public_` prefix.
        map.insert(
            "RecipientOwnerHash".to_string(),
            vec![bytes_to_decimal_string(&self.recipient_owner_hash)],
        );
        map.insert("MaxPrice".to_string(), vec![self.max_price.to_string()]);
        map.insert("CreatedAt".to_string(), vec![self.created_at.to_string()]);
        map.insert(
            "OrderAmount".to_string(),
            vec![self.order_amount.to_string()],
        );
        map.insert(
            "ExternalDataHash".to_string(),
            vec![bytes_to_decimal_string(&self.external_data_hash)],
        );
        map.insert(
            "PrivateTxBlinding".to_string(),
            vec![bytes_to_decimal_string(&self.private_tx_blinding)],
        );
        for (key, value) in utxo_witness_entries(&self.order_in, "OrderIn")
            .into_iter()
            .chain(utxo_witness_entries(&self.reservation_in, "ReservationIn"))
            .chain(utxo_witness_entries(&self.recipient_out, "RecipientOut"))
            .chain(utxo_witness_entries(&self.maker_counter, "MakerCounter"))
            .chain(utxo_witness_entries(&self.maker_source, "MakerSource"))
        {
            map.insert(key, value);
        }
        map
    }

    pub fn prove(&self) -> Result<OrderProof, ProofError> {
        negate_and_compress_proof(&ffi::prove(CircuitId::EscrowSettle, &self.witness())?)
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashSet;

    use super::*;
    use crate::utxo::expected_utxo_witness_keys;

    fn sample() -> EscrowSettleProofInputs {
        EscrowSettleProofInputs {
            public_input_hash: [1; 32],
            private_tx_hash: [2; 32],
            first_nullifier: [10; 32],
            execution_price: 90,
            max_price: 100,
            created_at: 1_700_000_000,
            order_in_hash: [3; 32],
            reservation_in_hash: [4; 32],
            recipient_owner_hash: [6; 32],
            authority_owner_hash: [7; 32],
            order_amount: 50,
            order_in: ProofInputUtxo::default(),
            reservation_in: ProofInputUtxo::default(),
            recipient_out: ProofInputUtxo::default(),
            maker_counter: ProofInputUtxo::default(),
            maker_source: ProofInputUtxo::default(),
            external_data_hash: [8; 32],
            private_tx_blinding: [9; 32],
        }
    }

    #[test]
    fn witness_key_set_matches_circuit_fields() {
        let witness = sample().witness();
        let keys: HashSet<&str> = witness.keys().map(String::as_str).collect();

        let mut expected: Vec<String> = vec![
            "Public_PublicInputHash".to_string(),
            "Public_PrivateTxHash".to_string(),
            "Public_FirstNullifier".to_string(),
            "Public_ExecutionPrice".to_string(),
            "Public_OrderInHash".to_string(),
            "Public_ReservationInHash".to_string(),
            "Public_AuthorityOwnerHash".to_string(),
            "RecipientOwnerHash".to_string(),
            "MaxPrice".to_string(),
            "CreatedAt".to_string(),
            "OrderAmount".to_string(),
            "ExternalDataHash".to_string(),
            "PrivateTxBlinding".to_string(),
        ];
        for prefix in [
            "OrderIn",
            "ReservationIn",
            "RecipientOut",
            "MakerCounter",
            "MakerSource",
        ] {
            expected.extend(expected_utxo_witness_keys(prefix));
        }

        let expected: HashSet<&str> = expected.iter().map(String::as_str).collect();
        assert_eq!(keys, expected);
    }
}
